js/web/lib/wasm/wasm-core-impl.ts

```diff
@@ -32,54 +32,72 @@ const activeSessions = new Map<number, SessionMetadata>();
+     wasm._OrtReleaseSessionOptions(sessionOptionsHandle);
+     allocs.forEach(wasm._free);
+   }
+
+   const inputCount = wasm._OrtGetInputCount(sessionHandle);
+   const outputCount = wasm._OrtGetOutputCount(sessionHandle);
+
+   const inputNames = [];
+   const inputNamesUTF8Encoded = [];
+   const outputNames = [];
+   const outputNamesUTF8Encoded = [];
+   for (let i = 0; i < inputCount; i++) {
+     const name = wasm._OrtGetInputName(sessionHandle, i);
+     if (name === 0) {
+       throw new Error('Can\'t get an input name');
+     }
+     inputNamesUTF8Encoded.push(name);
+     inputNames.push(wasm.UTF8ToString(name));
+   }
+   for (let i = 0; i < outputCount; i++) {
+     const name = wasm._OrtGetOutputName(sessionHandle, i);
+     if (name === 0) {
+       throw new Error('Can\'t get an output name');
+     }
+     outputNamesUTF8Encoded.push(name);
+     outputNames.push(wasm.UTF8ToString(name));
+   }
+
+   activeSessions.set(sessionHandle, [sessionHandle, inputNamesUTF8Encoded, outputNamesUTF8Encoded]);
+   return [sessionHandle, inputNames, outputNames];
+ };
 
-        wasm.HEAPU8.set(model, modelDataOffset);
-        sessionHandle = wasm._OrtCreateSession(modelDataOffset, model.byteLength, sessionOptionsHandle);
-        if (sessionHandle === 0) {
-          throw new Error('Can\'t create a session');
-        }
-      } finally {
-        wasm._free(modelDataOffset);
-        wasm._OrtReleaseSessionOptions(sessionOptionsHandle);
-        allocs.forEach(wasm._free);
-      }
 
-      const inputCount = wasm._OrtGetInputCount(sessionHandle);
-      const outputCount = wasm._OrtGetOutputCount(sessionHandle);
-
-      const inputNames = [];
-      const inputNamesUTF8Encoded = [];
-      const outputNames = [];
-      const outputNamesUTF8Encoded = [];
-      for (let i = 0; i < inputCount; i++) {
-        const name = wasm._OrtGetInputName(sessionHandle, i);
-        if (name === 0) {
-          throw new Error('Can\'t get an input name');
-        }
-        inputNamesUTF8Encoded.push(name);
-        inputNames.push(wasm.UTF8ToString(name));
-      }
-      for (let i = 0; i < outputCount; i++) {
-        const name = wasm._OrtGetOutputName(sessionHandle, i);
-        if (name === 0) {
-          throw new Error('Can\'t get an output name');
-        }
-        outputNamesUTF8Encoded.push(name);
-        outputNames.push(wasm.UTF8ToString(name));
-      }
-
-      activeSessions.set(sessionHandle, [sessionHandle, inputNamesUTF8Encoded, outputNamesUTF8Encoded]);
-      return [sessionHandle, inputNames, outputNames];
+/**
+ * create an instance of InferenceSession.
+ * @returns the metadata of InferenceSession. 0-value handle for failure.
+ */
+export const createSession =
+    (model: Uint8Array, options?: InferenceSession.SessionOptions): SerializableSessionMetadata => {
+      const modelData: SerializableModeldata = createSessionAllocate(model);
+      return createSessionFinalize(modelData, options);
```

Comment:
I think this cannot work because `model: Uint8Array` is still available in the context when calling `createSessionFinalize`

Reply:
need the following things happen inside a function:
- fetch and get the ArrayBuffer
- create Uint8Array from the ArrayBuffer
- allocate in wasm memory
- memcpy the Uint8Array into the allocated memory

then, when after calling this function, the JS ArrayBuffer is garenteed to be collectable. If passing the Uint8Array as parameter, this could not work

Reply:
understood. createSession handles the case where the user passes in a Uint8Array. 
The split only works when the users passes in some url and then we execute this code:
https://github.com/microsoft/onnxruntime/blob/2be734d0e7da32efcf7463d508954ccab272a1d4/js/web/lib/wasm/session-handler.ts#L60
which will call this function that should do what you are saying:
https://github.com/microsoft/onnxruntime/blob/2be734d0e7da32efcf7463d508954ccab272a1d4/js/web/lib/wasm/session-handler.ts#L41